docs/protocols/java_object_graph_spec.md

```diff
@@ -64,24 +83,62 @@ If schema consistent mode is enabled globally or enabled for current class, clas
 
 If schema evolution mode is enabled globally or enabled for current class, class meta will be written as follows:
 
-- If meta share mode is not enabled, class meta will be written as scheme consistent mode, field meta such as field type
+- If meta share mode is not enabled, class meta will be written as scheme consistent mode. Additionally, field meta such
+  as field type
   and name will be written when the object value is being serialized using a key-value like layout.
-- If meta share mode is enabled, class will be written as a unsigned int.
+- If meta share mode is enabled, class meta will be written as a meta-share encoded binary if class hasn't been written
+  before, otherwise unsigned it will be written as an unsigned varint which references to previous written class meta.
 
 ## Meta share
+
 > This mode will forbid streaming writing since it needs to look back for update the offset after the whole object graph
 > writing and mete collecting is finished.
-> TODO: We have plan to streamline meta writing but not started yet.
+> We plan to streamline meta writing but haven't started yet.
 
 ### Schema consistent
 
+Class will be encoded as an enumerated string by full class name.
 
 ### Schema evolution
 
+Class meta format:
+
+```
+| meta header: hash + num classes | current class meta | parent class meta | ... |
+```
+
+#### Meta header
+
+Meta header is a 64 bits number value encoded in little endian order.
+
+- Lowest 4 digits `0b0000~0b1110` are used to record num classes. `0b1111` is preserved to indicate that Fury need to
```

Comment:
max number of classes is 14?

Reply:
If number of classes  exceeds 14, `0b1111` will be used and another byte will be write next to indicate num classes

Reply:
Ahh I see.